Request body streaming

pwcsquared opened the pull request.

pwcsquared committed:
add support for stream body in HTTPoison request

pwcsquared committed:
upload bitstream with put_block requests for each chunk

pwcsquared committed:
enumerate transformed blob stream

temporary inspect added

pwcsquared committed:
fix default argument declarations for put_blob

pwcsquared committed:
fix syntax error

pwcsquared committed:
pass params to fn when committing block list

pwcsquared committed:
pass container and name params to commit block list

pwcsquared committed:
remove application/octet-stream content_type

pwcsquared committed:
remove x-ms-blob-type header from put block request

pwcsquared committed:
remove x-ms-blob-type header from commit blob list request

pwcsquared committed:
fix params arg formatting

pwcsquared committed:
hardcode content-type header and option for put block request

pwcsquared committed:
fix blockid param for put block request

pwcsquared committed:
base64 encode blockid param

pwcsquared committed:
inspect request result

pwcsquared committed:
add xml content-type header to commit block list request

pwcsquared committed:
add xml version, encoding tag to commit block list request body

pwcsquared committed:
switch content_type to plain text for commit block list request

pwcsquared committed:
fix typo in xml request body

pwcsquared committed:
reverse block id list on commit, URI encode blockid string

pwcsquared committed:
only URI encode blockid for signed key

HTTPoison will automatically URI encode params, and we want to avoid
doubly encoding blockid strings

pwcsquared committed:
undo blockid list reverse

pwcsquared committed:
revert URI encoding of formatted params

pwcsquared committed:
add content-length header for put_block request

pwcsquared committed:
syntax fix

pwcsquared committed:
make shared key authorized request with 2023-01-03 version

pwcsquared committed:
pad block_id to 64 character length

pwcsquared committed:
reverse block_id list for commit block list request

pwcsquared committed:
define function for copying blobs

pwcsquared committed:
add content-type header to copy_blob request

pwcsquared committed:
fix x-ms-copy-source header in copy_blob request

pwcsquared committed:
add content-type header to copy_blob request

pwcsquared committed:
move block operations to own module

pwcsquared committed:
fix typespec for nillable content_type param

mattpolzin commented on lib/azurex/blob.ex:
```diff
@@ -44,10 +45,34 @@ defmodule Azurex.Blob do
+    |> Stream.run()
+  end
+
+  def put_blob(name, blob, content_type, container, params) do
```
Does this for sure work with a nil content type?

pwcsquared replied:
Good catch, uploading a blob as an entire binary didn't previously work with `nil` for content type. I've gone back and added support for it so that the blob will default to the `application/octet-stream` content type.

mattpolzin commented on lib/azurex/blob.ex:
```diff
@@ -44,10 +45,34 @@ defmodule Azurex.Blob do
-  def put_blob(name, blob, content_type, container \\ nil, params \\ []) do
+  def put_blob(name, blob, content_type, container \\ nil, params \\ [])
+
+  def put_blob(name, {:stream, bitstream}, _content_type, container, params) do
```
Is there any drawback to not setting the content type when (or after) streaming an asset?

pwcsquared replied:
I don't believe so. Azure will set the content-type to `application/octet-stream` automatically if no type is specified when the block list is committed, which is our current behavior with MinIO. Though for the purposes of an open-source library, I should probably at least make it possible to provide a content-type if another user of this library wants that behavior

mattpolzin commented on lib/azurex/blob/block.ex:
```diff
@@ -0,0 +1,97 @@
+defmodule Azurex.Blob.Block do
+  @moduledoc """
+  Implementation of Azure Blob Storage
```
nit: could be more specific.

mattpolzin reviewed (commented):
Looks like a solid addition. What are your plans/timing for this draft relative to the asset service work that you've built on top of it?

pwcsquared committed:
move test dirs to match lib structure

pwcsquared committed:
create test containers in setup, add integration test for streaming blob

pwcsquared committed:
update Authorization header test expectations for newer x-ms-version

pwcsquared committed:
create test setup helper module for setting env, creating test containers and blob

pwcsquared committed:
add additional documentation detail for blob/block operations

pwcsquared committed:
add support for specifying content type for streaming blob body

also added support for `nil` content type in uploading blob as binary.
This will result in the uploaded blob being assigned the default conent
type "application/octet-stream".

pwcsquared committed:
mix format

pwcsquared committed:
add integration test for copy_blob/3

mattpolzin commented:
This looks great. I give it a thumbs up for Opal's usage -- and I pose to you: do you want to re-open this PR as-is against the upstream repository? certainly could do that, or could break it out so that copying and streaming to blobs were two distinct PRs. there's really no necessity to do it one way or the other because it can always be done in response to a request from the upstream maintainer anyway.

pwcsquared committed:
Merge branch 'master' into request-body-streaming

pwcsquared committed:
add example bitstream to put_blob function documentation

pwcsquared committed:
fix generated blockid to be only 32 characters long

mattpolzin commented:
This was upstreamed right?

mattpolzin commented:
oh that was interesting. I actually _didn't_ merge this PR into `master` just now like GitHub indicates. what I did was use the "sync fork" button to sync our `master` branch with the upstream one. GitHub I guess noticed that the result of that was this branch being a part of `master`.
